### PR TITLE
WIP - Add an eslint rule for detecting package level side effects

### LINKED ISSUE #13910: Add `sideEffects` property to package.json for our packages
**Description**
The `sideEffects` property indicates whether a package introduces any side effects. It can be specified as either `false` or as an array of files that have side effects.

By specifying this property, webpack is able to reduce the size of JavaScript bundles by eliminating unused imports:
https://webpack.js.org/guides/tree-shaking/

`sideEffects` is particularly relevant when implementing feature flags. Without it, features need to be flagged at every possible point. For further details, see this discussion: https://github.com/WordPress/gutenberg/pull/13829#discussion_r255878995
 
**How to implement this**
Packages will need to be assessed individually. The biggest cause of side effects is likely to be `registerStore`. Any function call in the outer-most scope can result in a side-effect, particularly if there's no return value:
```
registerStore( //... );
registerBlock( // ... );
```

Mutation of global or imported variables is another example:
```
import store from '@wordpress/core-data';

modifyStore( store );
store.foo = bar;

window.shim = window.shim || function() {
    // ...
};
```

Another good indication is any import that doesn't define any new variables. The imported file will almost certainly have side-effects:
```
import 'something-with-side-effects';
```

**Other ideas**
It might be possible to use a linting rule to enforce this. A quick look at npm shows limited prior art, so a custom rule might be needed.

### PR DESCRIPTION
**Pull request by talldan**
Fixes #13910

## Description
This checks for violations of the package.json `sideEffects` property. At the moment there are a couple of particular patterns it checks for:

- Function calls in the top-most scope where the return value isn't used. e.g.:
```
registerStore( // ... );
window.addEventListener( // ... );
```

- Imports of other modules that don't create any new variables. e.g.:
```
import './store';
```

When it detects these patterns and the file has not been added to the package.json `sideEffects` property it warns the developer to do so.

Other patterns that I've thought about:
- Mutation of an imported value. I think this is possible to add a lint rule for. I imagine there aren't (m)any instances of this, since I'd expect it to be considered bad practice and caught in a code review.
- Any others?

Patterns that are hard to lint for:
- If a mutation or side-effect happens as the result of a function call - e.g. `function initializeModule() { window.addEventListener( // ... ); }`, it's hard to catch since I'd have to trace back to the callsite.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/package-side-effects.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+Examples of **correct** code for this rule:
+
+`src/index.js`
+```js
+// Module imports can introduce side effects.
+import 'module-with-side-effects';
+
+// Function calls can introduce side effects when the return value is not used.
+window.addEventListener( 'resize', handleResize );
+registerStore( store );
```
Comment (noisysocks) [inline packages/eslint-plugin/docs/rules/package-side-effects.md]: This is the same example as above.

**Review by noisysocks — COMMENTED** [review]
Very interesting!! 🤯

File: packages/eslint-plugin/docs/rules/package-side-effects.md (hunk 1 above)
Comment (talldan) [reply packages/eslint-plugin/docs/rules/package-side-effects.md]: Yep, was trying to illustrate that adding the file to the sideEffects property resolves the situation, but I need to make it clearer. Will rewrite.

**Comment by gziolo** [comment]
It might be helpful to have this rule to catch which packages have side-effects but I don't expect this would be bullet-proof. I think we have maybe a handful of packages which don't have side-effects and they are very low-level like:
- `autop`
- `block-serialization-default-parser`
- `block-serialization-spec-parser`
- `compose`
- `deprecated`
- `element`
- `escape-html`
- `is-shallow-equal`
- `keycodes`
- `priority-queue`
- `redux-routine`
- `tokens-list`
- `url`
- `wordcount`

and maybe a few more which I didn't explore in depth. For sure this PR contains a few false positives like `data` or `api-fetch` which run some logic when imported.

**Comment by talldan** [comment]
@gziolo Thanks for taking a look. Do you have an example of where `data` and `api-fetch` introduce side effects?

I'm looking for other patterns that could be added to this rule, but as you say it'd be hard to make anything bullet-proof.

I think if a rule can help manage those side-effects it'd still be beneficial to making it a completely manual process, and should also help alert devs to the need to think about side effects.

**Comment by gziolo** [comment]
> @gziolo Thanks for taking a look. Do you have an example of where `data` and `api-fetch` introduce side effects?

Sure, a related line for `data`:
* https://github.com/WordPress/gutenberg/blob/master/packages/data/src/default-registry.js#L6

I might be wrong about `api-fetch` - it probably sets up all middlewares on the server in WordPress core.

**Comment by talldan** [comment]
Closing this as it has lain dormant for some time and there's a bit more thought that needs to go into it.

**Comment by gziolo** [comment]
> Closing this as it has lain dormant for some time and there's a bit more thought that needs to go into it.

Do you still think, we should we at least pick some of the packages manually and apply this magic field to those which are all pure?

**Pull request closed without merging**